File: src/main/java/at/searles/parsing/combinators/ReducerOpt.java
package at.searles.parsing.combinators;

import at.searles.parsing.*;
import at.searles.parsing.printing.ConcreteSyntaxTree;
import at.searles.parsing.printing.PartialConcreteSyntaxTree;
import org.jetbrains.annotations.NotNull;

public class ReducerOpt<T> implements Reducer<T, T>, Recognizable.Opt {
    private final Reducer<T, T> parent;

    public ReducerOpt(Reducer<T, T> parent) {
        this.parent = parent;
    }

    @Override
    public Recognizable parent() {
        return parent;
    }

    @Override
    public boolean recognize(ParserCallBack env, ParserStream stream) {
        return true;
    }

    @Override
    public T parse(ParserCallBack env, ParserStream stream, @NotNull T left) {
        long preStart = stream.start();

        T nonOptResult = parent.parse(env, stream, left);

        assert stream.start() == preStart;

        return nonOptResult != null ? nonOptResult : left;
    }

    @Override
    public PartialConcreteSyntaxTree<T> print(PrinterCallBack env, @NotNull T t) {
        PartialConcreteSyntaxTree<T> output = parent.print(env, t);
        return output != null ? output : new PartialConcreteSyntaxTree<>(t, ConcreteSyntaxTree.empty());
    }

    @Override
    public String toString() {
        return createString();
    }
}

File: src/main/java/at/searles/parsing/combinators/ReducerOpt.kt
package at.searles.parsing.combinators

import at.searles.parsing.ParserStream
import at.searles.parsing.Recognizer
import at.searles.parsing.Reducer
import at.searles.parsing.printing.EmptyTree
import at.searles.parsing.printing.PartialTree

class ReducerOpt<T>(private val parent: Reducer<T, T>) : Reducer<T, T> {

    override fun reduce(left: T, stream: ParserStream): T? {
        return stream.reduce(left, parent) ?: left
    }

    override fun recognize(stream: ParserStream): Boolean {
        stream.recognize(parent, false)
        return true
    }

    override fun print(item: T): PartialTree<T> {
        val output = parent.print(item)
        return output ?: PartialTree(item, EmptyTree)
    }

    override fun <V> plus(right: Reducer<T, V>): Reducer<T, V> {
        return ReducerOrReducer(parent + right, right)
    }

    override fun plus(right: Recognizer): Reducer<T, T> {
        return ReducerOrReducer(parent + right, right.toReducer())
    }

    override fun toString(): String {
        return "$parent.opt"
    }
}